projects/packages/jetpack-mu-wpcom/src/features/wpcom-block-editor/class-jetpack-wpcom-block-editor.php

```diff
@@ -0,0 +1,630 @@
+		global $pagenow;
+
+		// phpcs:ignore WordPress.Security.NonceVerification
+		return ( 'post.php' === $pagenow || 'post-new.php' === $pagenow ) && ! empty( $_GET['frame-nonce'] );
+	}
+
+	/**
+	 * Prevents frame options header from firing if this is a allowed iframe request.
+	 */
+	public function disable_send_frame_options_header() {
+		// phpcs:ignore WordPress.Security.NonceVerification, WordPress.Security.ValidatedSanitizedInput
+		if ( isset( $_GET['frame-nonce'] ) && $this->framing_allowed( $_GET['frame-nonce'] ) ) {
+			remove_action( 'admin_init', 'send_frame_options_header' );
+		}
+	}
+
+	/**
+	 * Adds custom admin body class if this is a allowed iframe request.
+	 *
+	 * @param string $classes Admin body classes.
+	 * @return string
+	 */
+	public function add_iframed_body_class( $classes ) {
+		// phpcs:ignore WordPress.Security.NonceVerification, WordPress.Security.ValidatedSanitizedInput
+		if ( isset( $_GET['frame-nonce'] ) && $this->framing_allowed( $_GET['frame-nonce'] ) ) {
+			$classes .= ' is-iframed ';
+		}
+
+		return $classes;
+	}
+
+	/**
+	 * Checks to see if cookie can be set in current context. If 3rd party cookie blocking
+	 * is enabled the editor can't load in iFrame, so emiting X-Frame-Options: DENY will
+	 * force the editor to break out of the iFrame.
+	 */
+	private function check_iframe_cookie_setting() {
+		if ( ! isset( $_SERVER['QUERY_STRING'] ) || ! strpos( filter_var( wp_unslash( $_SERVER['QUERY_STRING'] ) ), 'calypsoify%3D1%26block-editor' ) || isset( $_COOKIE['wordpress_test_cookie'] ) ) {
+			return;
+		}
+
+		if ( isset( $_SERVER['REQUEST_URI'] ) && empty( $_GET['calypsoify_cookie_check'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			header( 'Location: ' . esc_url_raw( wp_unslash( $_SERVER['REQUEST_URI'] ) . '&calypsoify_cookie_check=true' ) );
+			exit;
+		}
+
+		header( 'X-Frame-Options: DENY' );
+		exit;
+	}
+
+	/**
+	 * Allows to iframe the login page if a user is logged out
+	 * while trying to access the block editor from wordpress.com.
+	 */
+	public function allow_block_editor_login() {
+		// phpcs:ignore WordPress.Security.NonceVerification
+		if ( empty( $_REQUEST['redirect_to'] ) ) {
+			return;
+		}
+		// phpcs:ignore WordPress.Security.NonceVerification
+		$redirect_to = esc_url_raw( wp_unslash( $_REQUEST['redirect_to'] ) );
+
+		$this->check_iframe_cookie_setting();
+
+		$query = wp_parse_url( urldecode( $redirect_to ), PHP_URL_QUERY );
+		$args  = wp_parse_args( $query );
+
+		// Check nonce and make sure this is a Gutenframe request.
+		if ( ! empty( $args['frame-nonce'] ) && $this->framing_allowed( $args['frame-nonce'] ) ) {
+
+			// If SSO is active, we'll let WordPress.com handle authentication...
+			if ( ( new Modules() )->is_active( 'sso' ) ) {
+				// ...but only if it's not an Atomic site. They already do that.
+				if ( ! ( new Host() )->is_woa_site() ) {
+					add_filter( 'jetpack_sso_bypass_login_forward_wpcom', '__return_true' );
+				}
```

Comment:
Do we want to keep this? We might be able to just leave the flow regarding sso not being active.